### DISCUSSION TITLE
Jetstream Object store permissions?

### DISCUSSION THREAD

**Discussion by moo-im-a-cow** [nats-io/nats-server#5161] [Q&A]
Hi.

How do i configure jetstream permissions? i cant find any documentation on this

specifically i need:
disable/enable listing object stores
disable/enable listing files stored in an object store
disable/enable reading files stored in an object store (is it possible to do per file, or just per store?)
disable/enable writing/deleting files stored in an object store  (is it possible to do per file, or just per store?)
disable/enable creating / deleting object stores

basically what i want to do is make userA have read/list access to a specific object store, and userB have read/write/list access to that object store. neither user should be able to create, list or delete object stores (and no access to any other features of jetstream)

so far i've worked out that `$JS.>` gives full access to object stores and other parts of jetstream, but i have no clue what the layout for the rest of it is.

I found info for jetstream KV, https://github.com/nats-io/nats-server/discussions/4639, but cant find anything for objectstores

thanks.

**Comment by bjorndm** [comment]
Looking at this: 
https://github.com/nats-io/nats-server/blob/9fa359acea5b710bed8a2c8aedd9ff8348e0dd85/server/jetstream_api.go#L338
It seems we have to use $OBJ in stead of $KV.

**Comment by Jarema** [comment]
Object Store uses the prefix `$O` for its operations.

To see check what subjects are used for what, see the ADR: https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-20.md

Unfortunately, right now the permissions are possible only down to per-bucket level.

Working with `$O.<bucket>` permissions should give you per-bucket control.

**Reply by bjorndm** [reply]
With this config there is an admin user who can do anything, bdm can write to the object store and rdm can read from it but not write to it. This is experimental, perhaps not optimal yet.

```
jetstream {
    store_dir: "."
    max_memory_store = -1
    max_file_store = -1
}

accounts {
  APP {
  	default_permissions {
  		sub {
  			allow : [ "_INBOX.>" ]
  		}
  	}
    users: [
      {
      	user: "bdm"
      	password: "p4ss"
      	permissions {
	    	 pub {
      		 	allow : [ "$JS.API.STREAM.INFO.OBJ_image", "$JS.API.STREAM.MSG.GET.OBJ_image", "$O.image.>", "$JS.API.STREAM.PURGE.OBJ_image" ]
      		 }
      		 sub {
      		 	allow : [ "$O.image.>", "_INBOX.>" ]
      		 }
      		 allow_responses {
      		 	max : 1
      		 }
      		 allow_responses {
      		 	max : 1
      		 }
      	}
      },
     {
      	user: "rdm"
      	password: "p4ss"
      	permissions {
      		 pub {
      		 	allow : [ "$JS.API.STREAM.INFO.OBJ_image", "$JS.API.STREAM.MSG.GET.OBJ_image", "$JS.API.STREAM.NAMES", "$JS.API.CONSUMER.CREATE.OBJ_image.*.$O.image.>",
      		 		"$JS.API.CONSUMER.DELETE.OBJ_image.> ]
      		 }
      		 sub {
      		 	allow : [ "$O.image.>", "_INBOX.>" ]
      		 }
      		 allow_responses {
      		 	max : 1
      		 }
      	}
      },
       {
      	user: "admin"
      	password: "p4ss"
      	permissions {
      		 pub {
      		 	allow : [ ">"]
      		 }
      		 sub {
      		 	allow : [ ">"]
      		 }
      		 allow_responses {
      		 	max : 1
      		 }
      	}
      },
    ]
 	jetstream: true
  }

  SYS {
	users: [
      {
      	user: root, password: p4ass
      }
    ]
  }
}

system_account: SYS
```
